x/clp/keeper_test.go

```diff
@@ -98,4 +75,72 @@ func TestCoolKeeperCreate(t *testing.T) {
 	err5 := keeper.create(ctx, addr1, ticker3, name3, reserveRatio3)
 	require.Error(t, err5)
 
+	//Test cannot create CLP for base token
+	err6 := keeper.create(ctx, addr1, baseTokenTicker, baseTokenName, reserveRatio)
+	require.Error(t, err6)
+}
+
+func TestCoolKeeperTradeBase(t *testing.T) {
+	clpKey := sdk.NewKVStoreKey("clpTestKey")
+	ctx := setupContext(clpKey)
+	keeper, _, bankKeeper, address := setupKeepers(clpKey, ctx)
+
+	ticker := "eth"
+	name := "ethereum"
+	reserveRatio := 1
+	tenRune := sdk.NewCoin("RUNE", 10)
+	twentyRune := sdk.NewCoin("RUNE", 20)
+
+	keeper.create(ctx, address, ticker, name, reserveRatio)
+
+	//Test happy path trading
+	bankKeeper.SetCoins(ctx, address, sdk.Coins{tenRune})
+	err1 := keeper.tradeBase(ctx, address, ticker, 10)
+	coins := bankKeeper.GetCoins(ctx, address)
+	ethAmount := coins.AmountOf("eth").Int64()
+	runeAmount := coins.AmountOf("RUNE").Int64()
+	require.Nil(t, err1)
+	require.Equal(t, ethAmount, int64(10))
+	require.Equal(t, runeAmount, int64(0))
+
+	//Test double trade
+	bankKeeper.SetCoins(ctx, address, sdk.Coins{twentyRune})
+	keeper.tradeBase(ctx, address, ticker, 10)
+	keeper.tradeBase(ctx, address, ticker, 10)
+	coins = bankKeeper.GetCoins(ctx, address)
+	ethAmount = coins.AmountOf("eth").Int64()
+	runeAmount = coins.AmountOf("RUNE").Int64()
+	require.Equal(t, ethAmount, int64(20))
+	require.Equal(t, runeAmount, int64(0))
+
+	//Test invalid trade with nonexistent clp
+	bankKeeper.SetCoins(ctx, address, sdk.Coins{tenRune})
+	err2 := keeper.tradeBase(ctx, address, "btc", 10)
+	require.Error(t, err2)
+	coins = bankKeeper.GetCoins(ctx, address)
+	btcAmount := coins.AmountOf("btc").Int64()
+	runeAmount = coins.AmountOf("RUNE").Int64()
+	require.Equal(t, btcAmount, int64(0))
+	require.Equal(t, runeAmount, int64(10))
+
+	// //Test invalid trade with too little rune
+	bankKeeper.SetCoins(ctx, address, sdk.Coins{tenRune})
+	err3 := keeper.tradeBase(ctx, address, ticker, int64(20))
+	require.Error(t, err3)
+	coins = bankKeeper.GetCoins(ctx, address)
+	ethAmount = coins.AmountOf("eth").Int64()
+	runeAmount = coins.AmountOf("RUNE").Int64()
+	require.Equal(t, ethAmount, int64(0))
+	require.Equal(t, runeAmount, int64(10))
+
+	// //Test invalid trade with negative rune
+	bankKeeper.SetCoins(ctx, address, sdk.Coins{tenRune})
+	err4 := keeper.tradeBase(ctx, address, ticker, int64(-20))
+	require.Error(t, err4)
+	coins = bankKeeper.GetCoins(ctx, address)
+	ethAmount = coins.AmountOf("eth").Int64()
+	runeAmount = coins.AmountOf("RUNE").Int64()
+	require.Equal(t, ethAmount, int64(0))
+	require.Equal(t, runeAmount, int64(10))
+
```

Comment:
Great tests! I think we should check that the CLP's internal coins have been adjusted too. What is your opinion?